File: packages/snow_draw_core/lib/draw/elements/core/element_hit_tester.dart
import 'package:meta/meta.dart';

import '../../models/element_state.dart';
import '../../types/draw_point.dart';
import '../../types/draw_rect.dart';

/// Hit testing interface for a single element type.
@immutable
abstract interface class ElementHitTester {
  const ElementHitTester();

  /// Returns true if [position] hits [element].
  bool hitTest({
    required ElementState element,
    required DrawPoint position,
    double tolerance = 0,
  });

  /// Returns the element bounds used for selection overlays.
  DrawRect getBounds(ElementState element);
}

File: packages/snow_draw_rust_engine/src/draw/elements/core/element_hit_tester.rs
#![allow(dead_code)]

pub use crate::draw::models::element_state::ElementState;
use crate::draw::types::draw_point::DrawPoint;
use crate::draw::types::draw_rect::DrawRect;

/// Element snapshot used by element-specific hit testers.
///
/// This mirrors Dart by using the canonical draw-model element state.

/// Hit testing interface for a single element type.
///
/// Mirrors the Dart `ElementHitTester` abstraction while providing an idiomatic
/// Rust API without default arguments.
pub trait ElementHitTester: Send + Sync {
    /// Returns true when `position` hits `element` with zero tolerance.
    fn hit_test(&self, element: &ElementState, position: DrawPoint) -> bool {
        self.hit_test_with_tolerance(element, position, 0.0)
    }

    /// Returns true when `position` hits `element` with `tolerance`.
    fn hit_test_with_tolerance(
        &self,
        element: &ElementState,
        position: DrawPoint,
        tolerance: f64,
    ) -> bool;

    /// Returns the element bounds used for selection overlays.
    fn get_bounds(&self, element: &ElementState) -> DrawRect;
}
